src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,757 @@
+	codec_context->framerate.den           = 1;
+	codec_context->pix_fmt                 = AV_PIX_FMT_YUV420P;
+	codec_context->sample_aspect_ratio.num = static_cast<int>(
+	        pixel_aspect_ratio.Num());
+	codec_context->sample_aspect_ratio.den = static_cast<int>(
+	        pixel_aspect_ratio.Denom());
+
+	if (avcodec_open2(codec_context, codec, nullptr) < 0) {
+		LOG_ERR("FFMPEG: Failed to open video context");
+		return false;
+	}
+
+	frame = av_frame_alloc();
+	if (!frame) {
+		LOG_ERR("FFMPEG: Failed to allocate video frame");
+		return false;
+	}
+	frame->width               = width;
+	frame->height              = height;
+	frame->format              = static_cast<int>(codec_context->pix_fmt);
+	frame->sample_aspect_ratio = codec_context->sample_aspect_ratio;
+	frame->pts                 = 0;
+	// 0 means auto-align based on current CPU
+	constexpr int memory_alignment = 0;
+	if (av_frame_get_buffer(frame, memory_alignment) < 0) {
+		LOG_ERR("FFMPEG: Failed to get video frame buffer");
+		return false;
+	}
+
+	// Initalise the re-scaler. Currently only used to convert to YuV color
+	// space.
+	constexpr SwsFilter* source_filter = nullptr;
+	constexpr SwsFilter* dest_filter   = nullptr;
+	constexpr double* parameters       = nullptr;
+	rescaler_contex                    = sws_getContext(width,
+                                         height,
+                                         map_pixel_format(pixel_format),
+                                         width,
+                                         height,
+                                         AV_PIX_FMT_YUV420P,
+                                         SWS_BILINEAR,
+                                         source_filter,
+                                         dest_filter,
+                                         parameters);
+	if (!rescaler_contex) {
+		LOG_ERR("FFMPEG: Failed to get swscaler context");
+		return false;
+	}
+
+	this->width              = width;
+	this->height             = height;
+	this->pixel_format       = pixel_format;
+	this->frames_per_second  = frames_per_second;
+	this->pixel_aspect_ratio = pixel_aspect_ratio;
+	return true;
+}
+
+void FfmpegVideoEncoder::Free()
+{
+	if (rescaler_contex) {
+		sws_freeContext(rescaler_contex);
+		rescaler_contex = nullptr;
+	}
+	if (frame) {
+		av_frame_free(&frame);
+		frame = nullptr;
+	}
+	if (codec_context) {
+		avcodec_free_context(&codec_context);
+		codec_context = nullptr;
+	}
+}
+
+static void send_packets_to_muxer(AVCodecContext* context, int stream_index,
+                                  RWQueue<AVPacket*>& queue)
+{
+	int packet_received;
+	do {
+		AVPacket* packet = av_packet_alloc();
+		if (!packet) {
+			LOG_ERR("FFMPEG: Failed to allocate audio packet");
+			return;
+		}
+		packet_received = avcodec_receive_packet(context, packet);
+		if (packet_received == 0) {
+			packet->stream_index = stream_index;
+			if (!queue.Enqueue(std::move(packet))) {
+				// Should not ever happen
```

Comment:
Suggest asserting, because the logical is 100% programmatically controlled (and expected). 
This way of something is broken, it will crash right away in debug builds and the backtrace will land here. 
In general, you can assert whenever "programmatically guaranteed to happen", and avoid the error handling from being part of the release code paths. 
(And eliminating the if-body here will eliminate the PVS use-after-move warning too).

Reply:
The assert also is a good communication mechanism to readers that more firmly states 100% expectations.. where as if/else tells me, "well, realistically the code can go down both pathways at runtime, and I'm not really sure so we need to handle both routes".

Reply:
Yeah, I like that better.  I'll do that.

Reply:
I pushed a commit to do this and now I get unused variable warning in release builds.  I extracted the return value of `Enqueue` then asserted on it but the compiler gets mad when the assert gets compiled out for release.

Reply:
You need to be creative 😎

Reply:
`[[maybe_unsed]]` seems to work to fix it.